### PR TITLE
Add clarification to ClientCall javadoc 

### PR DESCRIPTION
**Pull request by groakley**
- Clarify that some methods of ClientCall may be invoked from any thread. 
- Adds @throws clauses to request(). 
- Adds threading note to cancel() to match request().

### CODE REVIEW

File: core/src/main/java/io/grpc/ClientCall.java [hunk 1]
```diff
@@ -205,6 +208,8 @@ public void onReady() {}
    * useful debug information. Both argument being null may log warnings and result in suboptimal
    * performance. Also note that the provided information will not be sent to the server.
    *
+   * <p>This method is safe to call from multiple threads without external synchronization.
```
Comment (carl-mastrangelo) [inline core/src/main/java/io/grpc/ClientCall.java]: I'm not sure this is threadsafe.  request causes events to be fired, but doesn't change the state of the object, where this does.  The only known implementation of ClientCall is not threadsafe for cancel (i.e. `ClientCallImpl`, which has unsafe access to `cancelCalled`).   @ejona86  can comment on whether this method _should_ be thread safe.

Comment (groakley) [reply core/src/main/java/io/grpc/ClientCall.java]: iiuc, cancel() is called directly when someone with a handle to a ListenableFuture (e.g. from a FutureStub) calls ListenableFuture.cancel(true). This could be on any thread. 

https://github.com/grpc/grpc-java/blob/master/stub/src/main/java/io/grpc/stub/ClientCalls.java#L483

If the current implementation's cancel() is not thread-safe, then I think it needs to be fixed.

Comment (carl-mastrangelo) [reply core/src/main/java/io/grpc/ClientCall.java]: That sounds plausible.    We probably didn't notice due to hardly ever using the ListenableFuture stubs.   In any case, the docs for `cancel()` cannot be updated before the implementation.

Comment (groakley) [reply core/src/main/java/io/grpc/ClientCall.java]: Makes sense. I will confirm this and then create a tracking issue if correct. In the meantime, removed comments about cancel().

File: core/src/main/java/io/grpc/ClientCall.java [hunk 2]
```diff
@@ -38,7 +38,8 @@
  * Applications are expected to utilize normal payload messages for such signals, as a response
  * naturally acknowledges its request.
  *
- * <p>Methods are guaranteed to be non-blocking. Implementations are not required to be thread-safe.
+ * <p>Methods are guaranteed to be non-blocking. Implementations are not required to be thread-safe
+ * except for {@link #request}, which may be called from any thread.
```
Comment (dapengzhang0) [inline core/src/main/java/io/grpc/ClientCall.java]: IMO Javadoc is facing the API users not the implementors, so `Implementations are not required ...` should not be mentioned here in the javadoc at all. It can be still be added in form of an ordinary comment `//...` or `/* ... */` out side the javadoc, or use the javadoc tag `@implNote` (require java 8).

Comment (dapengzhang0) [reply core/src/main/java/io/grpc/ClientCall.java]: Or rephrase it to be user-facing like `Not thread-safe except for  {@link #request}, which may be called from any thread.`

Comment (groakley) [reply core/src/main/java/io/grpc/ClientCall.java]: Reprhased. I did not clean up other examples of this kind of phrasing in this file, though.

File: core/src/main/java/io/grpc/ClientCall.java (hunk 1 above)
Comment (groakley) [reply core/src/main/java/io/grpc/ClientCall.java]: Created #4547 to track.
